pkg/kubelet/metrics/collectors/cri_metrics.go

```diff
@@ -0,0 +1,123 @@
+	"fmt"
+	"time"
+
+	"k8s.io/component-base/metrics"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+)
+
+type criMetricsCollector struct {
+	metrics.BaseStableCollector
+	descriptors           map[string]*metrics.Desc
+	listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error)
+}
+
+// Check if criMetricsCollector implements necessary interface
+var _ metrics.StableCollector = &criMetricsCollector{}
+
+// NewCRIMetricsCollector implements the metrics.Collector interface
+func NewCRIMetricsCollector(ctx context.Context, listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error), listMetricDescriptors func(context.Context) ([]*runtimeapi.MetricDescriptor, error)) metrics.StableCollector {
+	descs, err := listMetricDescriptors(ctx)
+	if err != nil {
+		klog.ErrorS(err, "Error reading MetricDescriptors")
+		return &criMetricsCollector{
+			listPodSandboxMetrics: listPodSandboxMetrics,
+		}
+	}
+	c := &criMetricsCollector{
+		listPodSandboxMetrics: listPodSandboxMetrics,
+		descriptors:           make(map[string]*metrics.Desc, len(descs)),
+	}
+
+	for _, desc := range descs {
+		c.descriptors[desc.Name] = criDescToProm(desc)
+	}
+
+	return c
+}
+
+// Describe implements the metrics.DescribeWithStability interface.
+func (c *criMetricsCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	for _, desc := range c.descriptors {
+		ch <- desc
+	}
+}
+
+// Collect implements the metrics.CollectWithStability interface.
+// TODO(haircommander): would it be better if these were processed async?
+func (c *criMetricsCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	podMetrics, err := c.listPodSandboxMetrics(context.Background())
+	if err != nil {
+		klog.ErrorS(err, "Failed to get pod metrics")
+		return
+	}
+
+	for _, podMetric := range podMetrics {
+		for _, metric := range podMetric.GetMetrics() {
+			promMetric, err := c.criMetricToProm(metric)
+			if err == nil {
+				ch <- promMetric
+			}
+		}
+		for _, ctrMetric := range podMetric.GetContainerMetrics() {
+			for _, metric := range ctrMetric.GetMetrics() {
+				promMetric, err := c.criMetricToProm(metric)
+				if err == nil {
+					ch <- promMetric
+				}
+			}
+		}
+	}
+}
+
+func criDescToProm(m *runtimeapi.MetricDescriptor) *metrics.Desc {
+	return metrics.NewDesc(m.Name, m.Help, m.LabelKeys, nil, metrics.INTERNAL, "")
+}
+
+func (c *criMetricsCollector) criMetricToProm(m *runtimeapi.Metric) (metrics.Metric, error) {
+	desc, ok := c.descriptors[m.Name]
+	if !ok {
+		err := fmt.Errorf("Error converting CRI Metric to prometheus format")
+		klog.V(5).ErrorS(err, "Descriptor not present in pre-populated list of descriptors", "descriptor name", m.Name)
+		return nil, err
+	}
+
+	typ := criTypeToProm[m.MetricType]
+
+	pm, err := metrics.NewConstMetric(desc, typ, float64(m.GetValue().Value), m.LabelValues...)
+	if err != nil {
+		klog.ErrorS(err, "Error getting summary for resourceMetric prometheus endpoint")
```

Comment:
If there is an error here, it means that prometheus was unable to unable intilize the constant metric (i.e. `prometheus.NewConstMetric` failed) right? Might be worth to include some more info then since it likely something may be wrong with the descriptor / values?: Maybe we can use the existing `desc.String()` function - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/desc.go#L91-L97

```
klog.ErrorS(err, "Error creating CRI prometheus metric", "metric", desc.String())
```